AuthorizedKeysCommand wrapper to read ~/.ssh/authorized_keys.d

bgilbert opened the pull request:
Implement https://github.com/coreos/fedora-coreos-tracker/issues/139 option 4.

cc @lucab @jlebon

jlebon commented on src/main.rs:
```diff
@@ -0,0 +1,116 @@
+    let opts = parse_args()?;
+
+    // switch user
+    let (user, _switch_guard) = switch_user(&opts.username)?;
```
Reading the docs on `switch_user_group`, it says that one should explicitly call `drop()` on the guard. Let's do that here before returning `Ok(())` so that if this code is ever refactored, we make sure that it comes along?

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+
+const KEYS_SUBDIR: &str = "authorized_keys.d";
+
+pub(crate) fn read_keys(ssh_dir: &Path, out: &mut dyn Write, err: &mut dyn Write) -> Result<()> {
```
Do we actually need dynamic dispatch here? Why not `impl Write`?

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+
+pub(crate) fn read_keys(ssh_dir: &Path, out: &mut dyn Write, err: &mut dyn Write) -> Result<()> {
+    // check permissions on paths leading to KEYS_SUBDIR
+    ensure_safe_dir_mode(ssh_dir)?;
```
Hmm, does this mean we error out if `~/.ssh` doesn't exist? Why is that an error instead of a no-op? (And similarly below for the authorized keys dir).

bgilbert replied:
I thought it was harmless, but it turns out that sshd does log non-zero exit status from an `AuthorizedKeysCommand`, so you're right.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+    let mut entries = read_dir(&authorized_keys_dir)
+        .chain_err(|| format!("reading {}", authorized_keys_dir.display()))?
+        .collect::<Vec<_>>();
+    entries.sort_unstable_by(|a, b| {
```
Huh, is ordering important for authorized keys?

bgilbert replied:
Yes.  If the same key is listed more than once, one e.g. with a forced-command and one without, the first instance will control.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+            )
+    });
+
+    // walk entries
```
Minor: redundant comment.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+    // unpack error
+    let ent = ent.chain_err(|| format!("reading {}", dir_path.display()))?;
+
+    // ignore dotfiles
+    if ent.file_name().into_vec()[0] == b'.' {
+        bail!("{} is a dotfile, ignoring", ent.path().display());
+    }
```
Hmm, seems it'd be cleaner to do the filtering upfront in the caller before sorting.

bgilbert replied:
sshd ignores stderr from `AuthorizedKeysCommand`s, so the detailed error reporting is only useful for manual debugging.  For that purpose, I think it's useful to retain the "is a dotfile, ignoring" message, and it's convenient to put that in `try_read_key_file()` because the caller already has infrastructure for logging errors to stderr and continuing.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+
+    // report results
+    let mut output: Vec<u8> = Vec::new();
+    // add comment with source path
```
Do these injected comments actually show up anywhere on disk? Or is it more to help with human testing?

bgilbert replied:
It's for testing.  The assumption is that the user may run the program manually to troubleshoot access problems.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,329 @@
+        .chain_err(|| format!("opening {}", ent.path().display()))?;
+    let mut contents: Vec<u8> = Vec::new();
+    file.read_to_end(&mut contents)
+        .chain_err(|| format!("reading {}", ent.path().display()))?;
```
Couldn't we just use `out.write_all()` for the comment on its own line and `io::copy` the actual file?

bgilbert commented:
- Updated for comments
- Defaulted the username argument to the current user
- Updated permissions checks to match sshd (mode & 0022 == 0 and owner in {root, $user})
- Reworked a bunch of tests

lucab commented:
The logic here seems reasonably small and isolate, and the overhead to maintain and package this as a standalone binary is non-negligible.
Should we consider sticking this logic as an Afterburn sub-command like `afterburn ssh-keydir` insted? It may not be the purest thing from a theoretic point of view, but I think the reduced maintenance burden largely outweighs that.

jlebon commented on src/keys.rs:
```diff
@@ -0,0 +1,430 @@
+
+fn copy_key_file_contents(path: &Path, file: &mut File, out: &mut impl Write) -> Result<()> {
+    let count = copy(file, out).chain_err(|| format!("copying {}", path.display()))?;
+    if count > 0 {
+        // non-empty file; ensure output is newline-terminated
+        file.seek(SeekFrom::Current(-1))
+            .chain_err(|| format!("seeking {}", path.display()))?;
+        let mut last: [u8; 1] = [0];
+        file.read(&mut last)
+            .chain_err(|| format!("reading last byte of {}", path.display()))?;
+        if last[0] != b'\n' {
+            out.write_all(&[b'\n']).chain_err(|| "writing newline")?;
+        }
+    }
```
Do we actually need all this? Why not just always write a newline?

jlebon commented:
>  Should we consider sticking this logic as an Afterburn sub-command like `afterburn ssh-keydir` insted? It may not be the purest thing from a theoretic point of view, but I think the reduced maintenance burden largely outweighs that.

Good point, worth considering. Another perhaps more underhanded alternative is to ship it as part of an Afterburn subpackage.

bgilbert commented:
Hmm.  Reasons not to ship as part of Afterburn:

- The tool is potentially useful on its own.  I could imagine someone installing it on Fedora Server or a non-Fedora distro.
- This is security-sensitive code.  I've tried to minimize dependencies and write the code carefully.  Shipping separately minimizes the risk of casual refactoring.
- The usual arguments against monorepos.
- As with [openssh-keys](https://github.com/coreos/openssh-keys) and [update-ssh-keys](https://github.com/coreos/update-ssh-keys), hopefully the code won't require much maintenance anyway.

bgilbert commented:
Updated.

lucab commented:
Ack, all fair points, no blocker from my side, I just wanted to explore alternatives routes.
